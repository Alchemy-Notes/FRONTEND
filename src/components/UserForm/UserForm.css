.form {
  display: grid;
  gap: 1rem;
  width: 400px;
  margin: 20px auto 10px;
  padding: 15px 20px 20px;
  text-align: left;
}

.form legend {
  text-align: center;
}

.formSection {
  display: grid;
  gap: 0.25rem;
}

.input {
  padding: 7px 5px;
  border: 1px solid var(--gray);
  font-size: 1.1rem;
  /* add a font-family */
}
