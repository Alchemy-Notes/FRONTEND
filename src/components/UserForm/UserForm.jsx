import React, { useState } from 'react';
import { useForm } from '../../hooks/useForm';
import styles from './UserForm.css';

export default function UserForm({
  className = '',
  label,
  //label = 'Authenticate,
  onSubmit,
  isSigningUp,
}) {
  const { formState, formError, handleFormChange, setFormError } = useForm({
    username: '',
    password: '',
  });
  const [loading, setLoading] = useState(false);

  const handleSubmit = async (e) => {
    e.preventDefault();
    const { username, password } = formState;

    try {
      if (!username || password.length < 8)
        throw new Error(
          'An username and password (with 8+ characters) are required.'
        );
      setLoading(true);
      await onSubmit(username, password);
    } catch (error) {
      setLoading(false);
      setFormError(error.message);
    }
  };

  return (
    <form className={className} onSubmit={handleSubmit}>
      <fieldset className={styles.form}>
        <legend>{label}</legend>
        <section className={styles.formSection}>
          <label htmlFor="username">Username </label>
          <input
            id="username"
            type="username"
            name="username"
            value={formState.username}
            className={styles.input}
            onChange={handleFormChange}
          />
        </section>
        <section className={styles.formSection}>
          <label htmlFor="password">Password </label>
          <input
            id="password"
            type="password"
            name="password"
            value={formState.password}
            className={styles.input}
            onChange={handleFormChange}
          />
        </section>
        <button type="submit" disabled={loading}>
          {loading ? 'Authenticating...' : label}
        </button>
        {formError && <p>{formError}</p>}
      </fieldset>
    </form>
  );
}
