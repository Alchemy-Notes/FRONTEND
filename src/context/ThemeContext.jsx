import { createContext, useContext, useState } from 'react';

const ThemeContext = createContext();

const ThemeProvider = ({ children }) => {
  const [theme, setTheme] = useState(false);

  return (
    <ThemeContext.Provider value={{ theme, setTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

const useTheme = () => {
  const context = useContext(ThemeContext);

  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeContext Provider');
  }

  return context;
};

export { ThemeProvider, useTheme };
