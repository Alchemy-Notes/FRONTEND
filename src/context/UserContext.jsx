import { createContext, useContext, useEffect, useState } from 'react';
import { getMe } from '../services/auth';
import { getUserNotes } from '../services/notes';

const UserContext = createContext();

const UserProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [notes, setNotes] = useState(null);
  const [tree, setTree] = useState(null);

  useEffect(() => {
    async function get() {
      const result = await getMe();
      setUser(result);
    }
    get();
  }, []);

  useEffect(() => {
    async function getNotes() {
      const result = await getUserNotes({
        userId: user.id,
        query: { type: 'favorite' },
      });
      setNotes(result);
    }
    if (user && user.id) getNotes();
  }, [user]);

  return user && notes ? (
    <UserContext.Provider
      value={{ user, setUser, notes, setNotes, tree, setTree }}
    >
      {children}
    </UserContext.Provider>
  ) : (
    <></>
  );
};

const useUser = () => {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('useUser must be used within a UserContext Provider');
  }

  return context;
};

export { UserProvider, useUser };
