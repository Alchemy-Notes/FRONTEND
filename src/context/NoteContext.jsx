import { createContext, useContext, useEffect, useState } from 'react';
import { getUserNotes } from '../services/notes';
import { useUser } from './UserContext';

const NoteContext = createContext();

const NoteProvider = ({ children }) => {
  const [notes, setNotes] = useState([]);
  const { user } = useUser();

  useEffect(() => {
    async function get() {
      const result = await getUserNotes({
        userId: user.id,
        query: { type: 'favorite' },
      });
      setNotes(result);
    }
    if (user && user.id) get();
  }, []);

  return (
    <NoteContext.Provider value={{ notes, setNotes }}>
      {children}
    </NoteContext.Provider>
  );
};

const useNotes = () => {
  const context = useContext(NoteContext);

  if (context === undefined) {
    throw new Error('useNotes must be used within a NoteContext Provider');
  }

  return context;
};

export { NoteProvider, useNotes };
