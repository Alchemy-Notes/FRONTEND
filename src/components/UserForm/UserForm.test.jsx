import { render } from '@testing-library/react';
import { UserProvider } from '../../context/UserContext';
import UserForm from './UserForm';

it('should display an empty user form for signing up/in', () => {
  const { container } = render(
    <UserProvider>
      <UserForm />
    </UserProvider>
  );

  expect(container).toMatchSnapshot();
});
