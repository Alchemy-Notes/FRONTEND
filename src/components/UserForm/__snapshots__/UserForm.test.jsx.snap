// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display an empty user form for signing up/in 1`] = `
<div>
  <form
    class=""
  >
    <fieldset
      class="form"
    >
      <section
        class="formSection"
      >
        <label
          for="username"
        >
          Username 
        </label>
        <input
          class="input"
          id="username"
          name="username"
          type="username"
          value=""
        />
      </section>
      <section
        class="formSection"
      >
        <label
          for="password"
        >
          Password 
        </label>
        <input
          class="input"
          id="password"
          name="password"
          type="password"
          value=""
        />
      </section>
      <button
        type="submit"
      />
      
    </fieldset>
  </form>
</div>
`;
